### PR TITLE
Dynamic Data Attribute RFC

### PR DESCRIPTION
**Pull request by harshjain2**
Dynamic Data Attribute RFC.

### CODE REVIEW

**Comment by AbhitejJohn** [comment]
General questions on this that do not seem to be coming out from the RFC:
1. How would this attribute be read by the adapter?
2. Would this be a hard coded attribute in the framework or would this use some level of extensibility? For instance after this effort, can someone have their own IDataDriven test which reads data over a stream?

**Comment by AbhitejJohn** [comment]
Tagging https://github.com/Microsoft/testfx/issues/141 for visibility.

**Comment by harshjain2** [comment]
`1. How would this attribute be read by the adapter?`
That's an implementation detail, right?

`2. Would this be a hard coded attribute in the framework or would this use some level of extensibility? For instance after this effort, can someone have their own IDataDriven test which reads data over a stream?`
Initial thought is to have another attribute `DynamicData` similar to attribute `DataRow`. 
It can be a good extensibility point if we can abstract out an interface and give users more control to implement  custom attributes for data driven tests.

File: RFCs/005-DynamicData-Attribute.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+### Requirements
+1. Test data declared as property should be easily asscoaited with test case.
+2. Test data declared as property should be associated with more than one test cases, if required.
+
+### Proposed solution
+Here is a solution that meets the above requirements:
+
+The property storing test data should be declared as static.
+```
+    static object[] ReusableTestData =
```
Comment (Faizan2304) [inline RFCs/005-DynamicData-Attribute.md]: Suggestion: Should we change it to IEnumerable<object[]>

**Comment by ipjohnson** [comment]
@harshjain2 I'm curious what's the relationship is between DynamicDataAttribute and DataRowAttribute? 

I'd be very interested in an extension point for this attribute. DataRow lacks the ability to provided data based on inspecting parameters of the method that is attributed

File: RFCs/005-DynamicData-Attribute.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## Motivation
+Often times, the inline test data need to be shared between tests. There could be more than one test cases in different files which will we using same test data. In most of these situations, user end up declaring same inline data for multiple test cases using "DataRow" attribute. Test framework should provides an attribute so that test data, declared as property, can be
+* Easily re-used
+* Easily organized
+
+## Detailed Design
+
+### Requirements
+1. Test data declared as property should be easily asscoaited with test case.
```
Comment (AbhitejJohn) [inline RFCs/005-DynamicData-Attribute.md]: should this actually be - "Inline test data can be reused by multiple test cases"? That would cover both the requirements you have.

File: RFCs/005-DynamicData-Attribute.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+```
+In a similary way, same test data can be associated with multiple test cases.
+
+### Benefits of using DynamicData attribute
+1. Leverages the familiar syntax of declaring inline data using DataRow attribute.
+2. More than one tests can use the same test data, if required. 
+3. Changes in the data can be scoped to single place for scenarios where test data is shared. 
+
+## Open questions
+1. Should we support methods and fields as well?
```
Comment (AbhitejJohn) [inline RFCs/005-DynamicData-Attribute.md]: Methods I believe would be a common scenario too.

File: RFCs/005-DynamicData-Attribute.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+### Proposed solution
+Here is a solution that meets the above requirements:
+
+The property storing test data should be declared as static.
+```
+    static object[] ReusableTestData =
+        {
+            new object[] {1, 2, 3},
+            new object[] {4, 5, 6}
+        };
```
Comment (AbhitejJohn) [inline RFCs/005-DynamicData-Attribute.md]: where is the return? I'm confused.

File: RFCs/005-DynamicData-Attribute.md (hunk 1 above)
Comment (JVimes) [inline RFCs/005-DynamicData-Attribute.md]: Would that force users to type more? I'd prefer to type even less (not repeat `object[]` for every row) by using 2d array:

    static object[,] ReusableTestData = new object[,]
    {
        {1,2,3},
        {4,5,6}
    };

DynamicData could support multiple input types. Ideas: `IEnumerable<object[]>`, `ICollection<object[]>`, `object[][]`, `object[,]`, `Action<...>`.

File: RFCs/005-DynamicData-Attribute.md (hunk 3 above)
Comment (JVimes) [inline RFCs/005-DynamicData-Attribute.md]: That would be powerful. And delegates would be good, too.

Comment (harshjain2) [inline RFCs/005-DynamicData-Attribute.md]: 👍 
Included methods as well. 
Delegates can be added as well, but it would be a bit difficult to include them in next release.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+Here is a solution using which test data can be easily assocaited with test case.
+
+The test framework should define an interface `IDataSource` which can be extended to get data from custom data source. The defintion of this interface is as below :
+```
+    public interface IDataSource
+    {
+        /// <summary>
+        /// Gets the data from custom data source.
+        /// </summary>
+        public IEnumerable<object[]> GetData();
```
Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Can you change this to `public IEnumerable<object[]> GetData(MethodInfo testMethod);` so that third party attributes can inspect the method before providing data. xUnit provides this and it's very handy.

File: RFCs/006-DynamicData-Attribute.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    {   new object[] {1, 2, 3},
+                        new object[] {4, 5, 6}
+                    }; 
+    }
+```
+
+This property can then be used to associate test data with data driven test case like below:
+```
+    [TestMethod]
+    [DynamicData("ReusableTestDataProperty")]
```
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: Can user use `nameof(MyPropertyData.ReusableTestDataProperty)`?

File: RFCs/006-DynamicData-Attribute.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    public void DynamicDataTestMethod1(int a, int b, int c)
+    {
+        Assert.IsTrue(1, a%3);
+        Assert.IsTrue(2, b%3);
+        Assert.IsTrue(3, c%3);
+    }
+
+    [TestMethod]
+    [DynamicData("ReusableTestDataMethod")]
+    public void DynamicDataTestMethod2(int a, int b, int c)
```
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: What could be the possible error scenarios:
1. Number of arguments mismatch
2. Type of args mismatch

How does the attribute work with generic types?

File: RFCs/006-DynamicData-Attribute.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+## Motivation
+Often times, the inline test data need to be shared between tests. There could be more than one test cases in different files which can use same test data. In most of these situations, user end up declaring same inline data for multiple test cases using "DataRow" attribute. Test framework should provide feature so that test data, declared as property, can be
+* Easily re-used
+* Easily organized
+
+## Detailed Design
+
+### Requirements
+1. Inline test data can be reused by multiple test cases.
```
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: 1. Support `Property` and `Method` available in current Type. It may be defined in a base type.
2. Support `Property` and `Method` given a fully qualified name?

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        public IEnumerable<object[]> GetData()
+        {
+            return new[]
+                    {   new object[] {1, 2, 3},
+                        new object[] {4, 5, 6}
+                    };
+        }
+    }
+
+    [TestMethod]
```
Comment (codito) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Shouldn't `DataTestMethod` understand these attributes?

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+##  Discovery of `IDataSource` attributes
+The MSTest v2 framework, on discovering a `TestMethod` probes additional attributes. On finding an attribute inheriting from `IDataSource`, framework invokes `GetData()` to fetch test data and iteratively invokes test method with the test data as arguments.
+
+## Benefits of using `IDataSource`
+1. Uses will be able to extend `IDataSource` to support custom data sources.
+2. More than one tests can re-use the test data defined in same data source.
+
+## Open questions
+1. Should we support multiple `IDataSource` attributes for a single test method.
```
Comment (codito) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Yes, please. It is up to the implementation. The framework shouldn't limit if user needs it.

Shouldn't `DataRow` implementation inherit from `IDataSource`?

File: RFCs/006-DynamicData-Attribute.md (hunk 8 above)
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: Can it load data source from a different assembly?

File: RFCs/006-DynamicData-Attribute.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+### Proposed solution
+Here is a solution that meets the above requirements:
+
+The property storing test data should be declared as static.
```
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: Is there a constraint on return type?

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        
+    }
+```
+In a similar way, multiple test methods can be decorated with same data source.
+
+##  Discovery of `IDataSource` attributes
+The MSTest v2 framework, on discovering a `TestMethod` probes additional attributes. On finding an attribute inheriting from `IDataSource`, framework invokes `GetData()` to fetch test data and iteratively invokes test method with the test data as arguments.
+
+## Benefits of using `IDataSource`
```
Comment (codito) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Have we looked at the `v1` data source attributes?

Should the data be available in `TestContext.DataRow` property?

File: RFCs/006-DynamicData-Attribute.md (hunk 11 above)
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: Some more thoughts:
I think `static` constraint may be limiting. Hypothetical scenario: say user is writing a test with one base test class and two inherited child classes testing two implementations. With static data, user can't override the datasource in the inherited children. There can't be a single test method with a single data source.

Is there an implementation challenge with non-static data source?

File: RFCs/006-DynamicData-Attribute.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
+# RFC 006- DynamicData Attribute for Data Driven Tests
+
+## Summary
+This details the MSTest V2 framework attribute "DynamicData" for data driven tests where test data, declared as properties or methods, can be shared between more than one test cases.
```
Comment (codito) [inline RFCs/006-DynamicData-Attribute.md]: Should this be called `DynamicDataRow` to keep it conceptually aligned with `DataRow`?

File: RFCs/006-DynamicData-Attribute.md (hunk 7 above)
Comment (ipjohnson) [inline RFCs/006-DynamicData-Attribute.md]: For type mismatch does it make sense to try and course the value using Converter.ChangeType or hard fail?

What happens if you don't supply values for parameters with defaults?

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 5 above)
Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: [Here](https://github.com/xunit/xunit/blob/master/src/xunit.core/Sdk/DataAttribute.cs) is the xunit attribute that I was thinking of.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 9 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Currently, `DataRow` can be decorated with both `DataTestMethod` and `TestMethod` attributes.

However, we should only support `DynamicData` with `DataTestMethod`. This could be a logical distinction between data-driven tests and normal tests.

From performance perspective also, this approach will be better as we need not to probe for `DynamicData` (or any other attribute inheriting from `IDataSource` while executing normal tests.

File: RFCs/006-DynamicData-Attribute.md (hunk 13 above)
Comment (harshjain2) [inline RFCs/006-DynamicData-Attribute.md]: >(thinking out aloud) Or is it that DataRow just indicates a single row, where as DynamicData represents list of rows?
To me, this sounds right. `DynamicData` can represent list of rows.

File: RFCs/006-DynamicData-Attribute.md (hunk 11 above)
Comment (harshjain2) [inline RFCs/006-DynamicData-Attribute.md]: If we use non-static data source, we basically need object of class in which Property or Method is declared. 
We need to ensure a default constructor exists for that class to be instantiated.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 12 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Having parameters is better that putting the data rows in TestContext.

In order to read data from TestContext, we need to apply explicit conversions.
```
[DynamicData("FooProperty"]
[DataTestMethod]
public void TestMethod1()
{
    string customerName = testContextInstance.DataRow["Name"].ToString();
    double balance = Convert.ToDouble(testContextInstance.DataRow["Balance"]);
    ...
}
```
The same can be implicitly achieved using parameters 
```
[DynamicData("FooProperty"]
[DataTestMethod]
public void TestMethod1(string customerName, double balance)
{
    ...
}
```

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 10 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: We should ideally do that, but this is out of scope for `DynamicData` related work.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 9 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Second thought, deprecate `[DataTestMethod]`. Let only `[TestMethod]` exist going forward.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 12 above)
Comment (JVimes) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: I second @harshjain2's vote for parameters. Much better syntax.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md [hunk 14]
```diff
@@ -1,58 +1,81 @@
...
 
 ### Proposed solution
-Here is a solution using which test data can be easily assocaited with test case.
+Here is a solution for using custom data source in data driven tests.
 
-The test framework should define an interface `IDataSource` which can be extended to get data from custom data source. The defintion of this interface is as below :
+The test framework should define an abstract class `TestDataSource` which can be extended to get data from custom data source.
 ```
-    public interface IDataSource
+    public abstract TestDataSource : Attribute
```
Comment (JVimes) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Is there a reason to have the word "Test" in the type name? I think it's implied by context.

Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Since it's an attribute shouldn't it end in Attribute? Though I question why the change from an interface to attribute, the interface seems to allow far more flexibility vs forcing inheritance?

File: RFCs/006-DynamicData-Attribute.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+## Summary
+This details the MSTest V2 framework attribute "DynamicData" for data driven tests where test data can be declared as properties or in methods and can be shared between more than one test cases.
+
+## Motivation
+Often times, data driven tests use shared test data which can be declared as properties or in methods. User can use `DataRow` for declaring inline data, but it can't be shared. Test framework should provide feature so that test data can declared as property or in method and can be easily used by multiple tests.
+
+## Detailed Design
+
+### Requirements
+1. Test data can be declared as properties or in methods and and can be by multiple test cases.
```
Comment (AbhitejJohn) [inline RFCs/006-DynamicData-Attribute.md]: nit: can be reused by multiple test cases

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 14 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: `Is there a reason to have the word "Test" in the type name? I think it's implied by context.`
It was renamed as we already a class [DataSource](https://github.com/Microsoft/testfx/blob/master/src/TestFramework/MSTest.Core/Attributes/VSTestAttributes.cs#L443).

Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: @ipjohnson Renamed it to `TestDataSourceAttribute`.

Comment (JVimes) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Users will wonder what the difference is. Why not overload the existing DataSource attribute instead of creating TestDataSource? I think these both do conceptually the same thing.

File: RFCs/006-DynamicData-Attribute.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
+# RFC 006- DynamicData Attribute for Data Driven Tests
+
+## Summary
+This details the MSTest V2 framework attribute "DynamicData" for data driven tests where test data can be declared as properties or in methods and can be shared between more than one test cases.
+
+## Motivation
+Often times, data driven tests use shared test data which can be declared as properties or in methods. User can use `DataRow` for declaring inline data, but it can't be shared. Test framework should provide feature so that test data can declared as property or in method and can be easily used by multiple tests.
```
Comment (jayaranigarg) [inline RFCs/006-DynamicData-Attribute.md]: test data can **be** declared as property

File: RFCs/006-DynamicData-Attribute.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+## Detailed Design
+
+### Requirements
+1. Test data can be declared as properties or in methods and and can be reused by multiple test cases.
+
+### Proposed solution
+Here is a solution that meets the above requirements:
+
+The static property or a static method having test data should be declared as below:
```
Comment (jayaranigarg) [inline RFCs/006-DynamicData-Attribute.md]: **_A_** static property

File: RFCs/006-DynamicData-Attribute.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    public void DynamicDataTestMethod2(int a, int b, int c)
+    {
+        Assert.AreEqual(1, a%3);
+        Assert.AreEqual(2, b%3);
+        Assert.AreEqual(0, c%3);
+    }
+}
+```
+
+In case, the property or method exists in a class other that the test class, an additional `Type` argument should be passed to `DynamicData` constructor.
```
Comment (jayaranigarg) [inline RFCs/006-DynamicData-Attribute.md]: exists in a class other **_then_** the test class

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 14 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: `Though I question why the change from an interface to attribute, the interface seems to allow far more flexibility vs forcing` 
I changed it from an interface to attribute as `TestDataSource` will always be used as an attribute. 
Also, we are providing a default implementation of `GetDisplayName()` method, making it an interface would enforce the users to always implement this method.

Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: @harshjain2 while I agree it will always be an attribute you're forcing the any 3rd party attribute to have to inherit from TestDataSourceAttribute. Where as if it's an interface 3rd party can make the choice on what to inherit from.

Using an interface doesn't stop you from providing an abstract with a GetDisplayName() implementation and it leaves the option of open for outside developers. 
```
public abstract class TestDataSourceAttribute : ITestDataSourceAttribute
{
 
}
```

Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: Do you have any use case for this scenario?

**Comment by harshjain2** [comment]
@ipjohnson @JVimes @jayaranigarg @AbhitejJohn @codito 
Do you guys think that `DynamicData` should also support non-static properties and non-static methods as well?

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 14 above)
Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: @harshjain2 I ran into this a while back when trying to combine xunit's datarow with another attribute and it occurred to me it would have been very useful if they had chosen an interface. 

I guess I'll ask what's the reason for not making it an interface? is there some technical challenge you're up against or do you just see forcing inheritance as an easier solution?

Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: From [implementation](https://github.com/Microsoft/testfx/pull/195/files#diff-34c5dff34694826180e3169be71418dfR26) perspective, this is how I am probing all the `TestDataSourceAttribute` decorated on a test method. 
Making it an interface, would add additional work of probing `Attribute` and then further deciding if it is inheriting from `ITestDataSource`.

Comment (ipjohnson) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: @harshjain2 I think you could probably get that done in one line of code by doing something like this

```
ITestDataSource[] dataSources = testMethod.GetAllAttributes(true).Where(a => a is ITestDataSource).OfType<ITestDataSource>().ToArray();
```
Then your code would works pretty the same as it is now. 

Making it an attribute is saying "No developer will ever want to make a data source attribute that inherits from anything other than TestDataSourceAttribute". Is this an assumption that is safe to make?

**Comment by ipjohnson** [comment]
@harshjain2 I don't personally see any use cases that an instance property would support that a static property can't support. Especially considering that there is no dependency injection support for test classes.

**Comment by JVimes** [comment]
I can't think of a need for instance data/methods.

File: RFCs/005-Framework-Extensibility-Custom-DataSource.md (hunk 14 above)
Comment (harshjain2) [inline RFCs/005-Framework-Extensibility-Custom-DataSource.md]: @ipjohnson I have made relevant changes, now ITestDataSource exists.

**Comment by JVimes** [comment]
Tried out DynamicData with a property. Worked great! Thanks so much. 😄 

I had one thought: it would be nice to use a flat array when there's only one parameter. But, only worth implementing if it is safe and easy.
